### PR TITLE
SAMZA-1361: OperatorImplGraph is using wrong keys to store/retrieve OperatorImpl in the map

### PR DESCRIPTION
**Pull request by prateekm**

### CODE REVIEW

**Comment by prateekm** [comment]
@nickpan47 @vjagadish1989 Please take a look.

File: samza-core/src/main/java/org/apache/samza/operators/StreamGraphImpl.java [hunk 1]
```diff
@@ -170,13 +170,13 @@ public ContextManager getContextManager() {
...
    *
-   * @return  a set of all available {@link OperatorSpec}s
+   * @return  all available {@link OperatorSpec}s
    */
   public Collection<OperatorSpec> getAllOperatorSpecs() {
     Collection<InputOperatorSpec> inputOperatorSpecs = inputOperators.values();
     Set<OperatorSpec> operatorSpecs = new HashSet<>();
-
     for (InputOperatorSpec inputOperatorSpec: inputOperatorSpecs) {
+      operatorSpecs.add(inputOperatorSpec);
```
Comment (vjagadish1989) [inline samza-core/src/main/java/org/apache/samza/operators/StreamGraphImpl.java]: thanks!

File: samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java [hunk 2]
```diff
@@ -129,7 +129,7 @@ public void close() {
    */
   OperatorImpl createAndRegisterOperatorImpl(OperatorSpec prevOperatorSpec, OperatorSpec operatorSpec,
       Config config, TaskContext context) {
-    if (!operatorImpls.containsKey(operatorSpec) || operatorSpec instanceof JoinOperatorSpec) {
+    if (!operatorImpls.containsKey(operatorSpec.getOpName()) || operatorSpec instanceof JoinOperatorSpec) {
```
Comment (nickpan47) [inline samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java]: This relies on the guarantee that getOpName() returns the unique ID for an operatorSpec in a user defined operator graph. Instead, using the operatorSpec object reference directly as key does not require that. Curious about what's the advantage here using String as the key?

Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java]: @nickpan47 There are 2 PartialJoinOperatorImpl instances corresponding to the same JoinOperatorSpec, which we need to distinguish between. Hence using their names (E.g. join-2-L and join-2-R) as the map keys.

Comment (nickpan47) [inline samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java]: Got it. As discussed offline, please make the requirements of the getOpName() return value from OperatorSpec clear s.t. implementation is required to return a unique ID string in a user defined DAG.

Comment (prateekm) [inline samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java]: @nickpan47 Updated documentation for that method, please take a look.

**Comment by nickpan47** [comment]
Note: there are checkstyleTest failures in the Jenkins build. Please fix those as well.

Thanks!

**Comment by prateekm** [comment]
@nickpan47 checkstyle failures were from an unrelated commit, should be fixed now. Will push a dummy commit to verify.

**Pull request closed without merging**